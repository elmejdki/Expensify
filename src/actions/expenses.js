import database from '../firebase/firebase';

export const addExpense = (expense = {}) => ({
  type: 'ADD_EXPENSE',
  expense
});

export const startAddExpense = (expenseData = {}) => {
  return (dispatch, getState) => {
    const uid = getState().auth.uid;

    const {
      description = "",
      note = '',
      amount = 0,
      createdAt = 0
    } = expenseData;

    const expense = { description, note, amount, createdAt };

    return database.ref(`users/${uid}/expenses`).push(expense).then((snapshot) => {
      dispatch(addExpense({
        id: snapshot.key,
        ...expense
      }));
    });
  }
}

export const removeExpense = (id) => ({
  type: 'REMOVE_EXPENSE',
  id
});

export const startRemoveExpense = (id) => {
  return (dispatch, getState) => {
    const uid = getState().auth.uid;

    return database.ref(`users/${uid}/expenses/${id}`).remove().then(() => {
      dispatch(removeExpense(id));
    });
  };
};

export const editExpense = (id, updates) => ({
  type: 'EDIT_EXPENSE',
  id,
  updates
});

export const startEditExpense = (id, updates) => {
  return (dispatch, getState) => {
    const uid = getState().auth.uid;

    return database.ref(`users/${uid}/expenses/${id}`).update(updates).then(() => {
      dispatch(editExpense(id, updates));
    });
  };
};

export const setExpenses = (expenses) => {
  return {
    type: 'SET_EXPENSES',
    expenses
  }
}

export const startSetExpenses = () => {
  return (disptach, getState) => {
    const uid = getState().auth.uid;

    return database.ref(`users/${uid}/expenses`).once('value').then((snapshot) => {
      const expenses = [];

      snapshot.forEach((childSnapshot) => {
        expenses.push({
          id: childSnapshot.key,
          ...childSnapshot.val()
        });
      });

      disptach(setExpenses(expenses));
    });
  };
};
