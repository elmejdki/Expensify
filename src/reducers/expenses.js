const expensesReducerDefaultState = [];

const expensesReducer = (state = expensesReducerDefaultState,
  action) => {
  switch(action.type) {
    case 'ADD_EXPENSE':
      return [
        ...state, action.expense
      ];
    case 'REMOVE_EXPENSE':
      return state.filter((expense) => expense.id !== action.id );
    case 'EDIT_EXPENSE':
      return state.map(expense => {
        if(expense.id === action.id) {
          return {
            ...expense,
            ...action.updates
          };
        }else {
          return expense;
        }
      });
    default:
      return state;
  };
}

export default expensesReducer;