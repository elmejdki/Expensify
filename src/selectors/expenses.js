const getVisibleExpenses = (expenses, { text, startDate, endDate,sortBy }) => {
  return expenses.filter(
    (expense) => {
      let {description, note} = expense;
      description = description.toLowerCase();
      note = note.toLowerCase();
      text = text.toLowerCase();

      return (
        description.includes(text) || note.includes(text)
      ) && (
        (
          typeof startDate !== 'number' || expense.createdAt >= startDate
        ) && (
          typeof endDate !== 'number' || expense.createdAt <= endDate
        )
      )
  }).sort((expense_a, expense_b) => 
    sortBy === 'amount' ?
      expense_b.amount - expense_a.amount :
      expense_b.createdAt - expense_a.createdAt
  );
}

export default getVisibleExpenses;