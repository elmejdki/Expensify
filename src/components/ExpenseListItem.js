import React from 'react';
import { connect } from 'react-redux';
import { removeExpense } from '../actions/expenses';

const ExpenseListItem = ({ id, description, amount, createdAt, removeExpense }) => (
  <div>
    <h3>{description}</h3>
    <p>
      {amount} - {
        new Date(createdAt)
          .toString()
          .match(/^\w{3}(\s*.)+\s\d{4}/)[0] }
    </p>
    <button
      onClick={() => {
        removeExpense(id); 
      }}
    >
      Remove
    </button>
  </div>
);

const mapDispatchToProps = { removeExpense };

export default connect(null, mapDispatchToProps)(ExpenseListItem);