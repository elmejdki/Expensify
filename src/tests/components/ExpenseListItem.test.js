import React from 'react';
import ShallowRenderer from 'react-test-renderer/shallow';
import ExpenseListItem from '../../components/ExpenseListItem';
import expenses from '../fixtures/expenses';

const renderer = new ShallowRenderer();

it('should render ExpenseListItem with the giving expense', () => {
  renderer.render(<ExpenseListItem {...expenses[0]} />);

  expect(renderer.getRenderOutput()).toMatchSnapshot();
});
