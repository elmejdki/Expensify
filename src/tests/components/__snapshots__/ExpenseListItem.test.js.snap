// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`should render ExpenseListItem with the giving expense 1`] = `
<Link
  className="list-item"
  to="/edit/1"
>
  <div>
    <h3
      className="list-item__title"
    >
      Gum
    </h3>
    <span
      className="list-item__sub-title"
    >
      1st January, 1970
    </span>
  </div>
  <h3
    className="list-item__data"
  >
    $1.95
  </h3>
</Link>
`;
